File: src/main/kotlin/com/linqingying/cangjie/psi/CjTypeParameterListOwnerNotStubbed.java
package com.linqingying.cangjie.psi;

import com.linqingying.cangjie.CjNodeTypes;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;


@Deprecated
abstract class CjTypeParameterListOwnerNotStubbed extends CjNamedDeclarationNotStubbed implements CjTypeParameterListOwner {
    public CjTypeParameterListOwnerNotStubbed(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    @Nullable
    public CjTypeParameterList getTypeParameterList() {
        return findChildByType(CjNodeTypes.TYPE_PARAMETER_LIST);
    }

    @Override
    @Nullable
    public CjTypeConstraintList getTypeConstraintList() {
        return findChildByType(CjNodeTypes.TYPE_CONSTRAINT_LIST);
    }

    @Override
    @NotNull
    public List<CjTypeConstraint> getTypeConstraints() {
        CjTypeConstraintList typeConstraintList = getTypeConstraintList();
        if (typeConstraintList == null) {
            return Collections.emptyList();
        }
        return typeConstraintList.getConstraints();
    }

    @Override
    @NotNull
    public List<CjTypeParameter> getTypeParameters() {
        CjTypeParameterList list = getTypeParameterList();
        if (list == null) return Collections.emptyList();

        return list.getParameters();
    }
}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjTypeParameterListOwnerNotStubbed.kt
package com.linqingying.cangjie.psi

import com.linqingying.cangjie.CjNodeTypes
import com.intellij.lang.ASTNode

@Deprecated("")
abstract class CjTypeParameterListOwnerNotStubbed(node: ASTNode) :
    CjNamedDeclarationNotStubbed(node), CjTypeParameterListOwner {
    override val typeParameterList: CjTypeParameterList?
        get() = findChildByType(CjNodeTypes.TYPE_PARAMETER_LIST)

    override val typeConstraintList: CjTypeConstraintList?
        get() = findChildByType(CjNodeTypes.TYPE_CONSTRAINT_LIST)

    override val typeConstraints: List<CjTypeConstraint>
        get() {
            val typeConstraintList = typeConstraintList ?: return emptyList()
            return typeConstraintList.constraints
        }

    override val typeParameters: List<CjTypeParameter>
        get() {
            val list = typeParameterList ?: return emptyList()

            return list.parameters
        }
}
